syntax = "proto3";

service ChatService {
  rpc Register (RegisterRequest) returns (RegisterResponse);
  rpc Login (LoginRequest) returns (LoginResponse);
  rpc ListUsers (ListUsersRequest) returns (ListUsersResponse);
  rpc SendMessage (SendMessageRequest) returns (SendMessageResponse);
  rpc Subscribe (SubscribeRequest) returns (stream Message);
  rpc MarkRead (MarkReadRequest) returns (MarkReadResponse);
  rpc DeleteUnreadMessage (DeleteUnreadMessageRequest) returns (DeleteUnreadMessageResponse);
  rpc ReceiveMessages (ReceiveMessagesRequest) returns (ReceiveMessagesResponse);
  rpc DeleteAccount (DeleteAccountRequest) returns (DeleteAccountResponse);
  rpc Logout (LogoutRequest) returns (LogoutResponse);
}

message RegisterRequest {
  string username = 1;
  string password = 2;
}

message RegisterResponse {
  string message = 1;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  string message = 1;
  int32 unread_messages = 2;
}

message ListUsersRequest {}
message ListUsersResponse {
  repeated string users = 1;
}

message SendMessageRequest {
  string sender = 1;
  string recipient = 2;
  string message = 3;
}

message SendMessageResponse {
  string message_id = 1;
}

message SubscribeRequest {
  string username = 1;
}

message Message {
  string id = 1;
  string sender = 2;
  string recipient = 3;
  string message = 4;
  string status = 5;
}

message MarkReadRequest {
  string username = 1;
  string contact = 2;
  int32 batch_num = 3;
}

message MarkReadResponse {
  string message = 1;
}

message DeleteUnreadMessageRequest {
  string sender = 1;
  string recipient = 2;
  string message_id = 3;
}

message DeleteUnreadMessageResponse {
  string message = 1;
}

message ReceiveMessagesRequest {
  string username = 1;
}

message ReceiveMessagesResponse {
  repeated Message messages = 1;
}

message DeleteAccountRequest {
  string username = 1;
}

message DeleteAccountResponse {
  string message = 1;
}

message LogoutRequest {
  string username = 1;
}

message LogoutResponse {
  string message = 1;
}
